src/Twine/Async.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Twine.Async (
+    AsyncError (..)
+  , renderAsyncError
+  , waitWithTimeout
+  , waitEitherBoth
+  ) where
+
+
+import           Control.Concurrent.Async (Async, waitSTM, waitEither)
+import           Control.Concurrent.Async (async, cancel, wait)
+import           Control.Concurrent.STM (atomically, orElse, retry)
+import           Control.Exception.Base (AsyncException (..))
+import           Control.Monad.Catch (catch, throwM)
+
+import           Control.Monad.IO.Class (liftIO)
+
+import           Data.Text (Text)
+import qualified Data.Text as T
+
+import           P
+
+import           Twine.Snooze
+
+import           System.IO (IO)
+
+import           X.Control.Monad.Trans.Either
+
+data AsyncError =
+  AsyncTimeout Duration
+  deriving (Eq, Show)
+
+renderAsyncError :: AsyncError -> Text
+renderAsyncError e =
+  case e of
+    AsyncTimeout d ->
+      mconcat [
+          "Async took greater than '"
+        , T.pack . show $ toSeconds d
+        , " seconds' to return."
+        ]
+
+waitWithTimeout :: Async a -> Duration -> EitherT AsyncError IO a
+waitWithTimeout a d = do
+  s <- liftIO . async $ snooze d
+  e <- liftIO $ waitEither a s
+  case e of
+    Left a' ->
+      pure $ a'
+    Right _ -> do
+      liftIO $ cancel a
+      (liftIO $ wait a)
+        `catch`
+          (\(ae :: AsyncException) ->
+            case ae of
+              ThreadKilled ->
+                left $ AsyncTimeout d
+              StackOverflow ->
```

Comment:
Dumb question - could this just be `e` at this point if you only care about `ThreadKilled`? Or do you want to be future safe?